### DISCUSSION TITLE
Explore cotton folders

### DISCUSSION THREAD

**Discussion by Anthony-DE-FARIA** [wrabit/django-cotton#125] [Q&A] [answered]
Hey !

I just discovered `django-cotton` and I'm very excited to start the implementation in my project! good job and thank you @wrabit I was exactly waiting for that.

I have a question about the setting.

I would like to keep my project architecture like so:

- `templates/app_1/cotton`
- `templates/app_2/cotton`
- `templates/app_3/cotton`

Instead of 
- `templates/app_1`
- `templates/cotton/app_1`
- `templates/app_2/`
- `templates/cotton/app_2`

Is it possible to allow cotton to explore?

Thanks :)

**Comment by wrabit** [comment]
Hi @Anthony-DE-FARIA !

It's an area I didn't specifically look at yet, someone else mentioned it recently too. I will put it higher on the list and get back here as soon as I've explored it.

**Comment by mattbha** [comment]
i'd like this, too

**Comment by wrabit** [comment]
@Anthony-DE-FARIA, @mattbha I'm taking a look at this and have a question

Say if we have:

```
templates/app_1/cotton/page_note.html
templates/app_2/cotton/tree.html
templates/app_3/cotton/frame.html
```

I have a concern by allowing component reference without app name that it will be easy to get confused as to where the component is located.

```
<c-page-note />
<c-tree />
<c-frame />
```

Is this a real concern or a non-issue for you?

As a side note, there are early plans to create vscode + pycharm plugin for component navigation which should minimise this concern.

**Reply by Anthony-DE-FARIA** [reply]
Hello @wrabit, I agree with you that this can be confusing.

Personally, I've found the HTML-like syntax of **dj-cotton** very convenient for organizing reusable components. To streamline our structure, I reorganized the project's template hierarchy, placing all reusable components within a single cotton folder. Fortunately, the project is still small, which made this reorganization feasible. I'm also considering replacing the classic {% include %} syntax with **dj-cotton** tags, as it enhances readability and provides greater control over component customization.

Here’s the template structure I currently use:

```
templates/app_1
templates/app_2
templates/cotton
```
However, I recognize that some teams prefer an alternative approach, like:

```
app_1/templates
app_2/templates
```
In my current setup, template files are organized as follows:

`templates/app_1/chart.html `– Currently using `{% include 'app_1/chart.html' with... %}`.
`templates/app_2/table.html `– Currently using `{% include 'app_2/table.html' with... %}`.

Reusable components are placed in the cotton directory, for example:
`templates/cotton/forms/inputs/currency.html `– Used as `<c-forms.inputs.currency />`.

For components like chart and table, I’d use a syntax such as:
`templates/cotton/app_1/chart.html` – `<c-app-1.chart />`.
`templates/cotton/app_2/table.html` – `<c-app-2.table />`.

This structure has been **highly effective**, as the component tags naturally include the app name and relevant subfolder, which provides clarity when reading the code.

However, there’s also the option to replicate app folders within the cotton directory, which may introduce some redundancy, like so:

`templates/app_1/example.html`
`templates/cotton/app_1/chart.html `-> `<c-app-1.chart />`

Alternatively, consolidating all cotton components within their respective app folders could reduce redundancy and keep everything centralized:

```
templates/app_1/example.html
templates/app_1/cotton/chart.html
```
As you pointed out, this approach does come with the potential downside of losing clarity, as it obscures the exact app structure within the cotton folder.

Ultimately, a plugin for VS Code or PyCharm to facilitate component navigation would be incredibly helpful in mitigating these challenges.

**Reply by mattbha** [reply]
> In my current setup, template files are organized as follows:
> 
> templates/app_1/chart.html – Currently using {% include 'app_1/chart.html' with... %}.
> templates/app_2/table.html – Currently using {% include 'app_2/table.html' with... %}.
> 
> Reusable components are placed in the cotton directory, for example:
> templates/cotton/forms/inputs/currency.html – Used as <c-forms.inputs.currency />.


I really like this structure, and would vote for it. My main project that uses cotton currently is effectively single-app, so I might yield my opinion overall on this issue.

**Reply by wrabit** [reply]
@Anthony-DE-FARIA thanks for the thorough response, I think it's clear what needs to be done. Current family commitments mean that I'm out for some days, but this will be done quite soon.

**Comment by wrabit** [comment]
@mattbha, @Anthony-DE-FARIA 

Another question..

Are there occasions when you want to achieve this pattern from any other directory other than from your project root’s templates folder?

(`os.path.join(BASE_DIR, 'templates')`)

**Reply by Anthony-DE-FARIA** [reply]
Not in my case but I know that's some people prefer to use the structure :
```
app_1/templates
app_2/templates
```

**Reply by mattbha** [reply]
I've done it both ways, I prefer globally used templates (such as cotton components) in the `BASE_DIR / "templates"`

**Comment by wrabit** [comment, marked as answer]
@Anthony-DE-FARIA @mattbha this is released in [1.4.0](https://github.com/wrabit/django-cotton/releases/tag/v1.4.0). 

Docs updated here https://django-cotton.com/docs/quickstart#templates-location

Hopefully this provides the capability we want.

**Reply by Anthony-DE-FARIA** [reply]
BRAVO! ☕️
